lib/STM_Core.ml

```diff
@@ -0,0 +1,281 @@
+    val check_obs : (Spec.cmd * res) list -> (Spec.cmd * res) list -> (Spec.cmd * res) list -> Spec.state -> bool
+    val gen_cmds_size : Spec.state -> int Gen.t -> Spec.cmd list Gen.t
+  (*val shrink_triple : ...*)
+    val arb_cmds_par : int -> int -> (Spec.cmd list * Spec.cmd list * Spec.cmd list) arbitrary
+    val all_interleavings_ok : Spec.cmd list -> Spec.cmd list -> Spec.cmd list -> Spec.state -> bool
+end
+=
+struct
+  (** {3 The resulting test framework derived from a state machine specification} *)
+
+  let rec gen_cmds s fuel =
+    Gen.(if fuel = 0
+         then return []
+         else
+  	  (Spec.arb_cmd s).gen >>= fun c ->
+	   (gen_cmds (Spec.next_state c s) (fuel-1)) >>= fun cs ->
+             return (c::cs))
+  (** A fueled command list generator.
+      Accepts a state parameter to enable state-dependent [cmd] generation. *)
+
+  let rec cmds_ok s cs = match cs with
+    | [] -> true
+    | c::cs ->
+      Spec.precond c s &&
+	let s' = Spec.next_state c s in
+	cmds_ok s' cs
+  (** A precondition checker (stops early, thanks to short-circuit Boolean evaluation).
+      Accepts the initial state and the command sequence as parameters.  *)
+
+  let arb_cmds s =
+    let cmds_gen = Gen.sized (gen_cmds s) in
+    let shrinker = match (Spec.arb_cmd s).shrink with
+                    | None   -> Shrink.list ~shrink:Shrink.nil (* no elem. shrinker provided *)
+                    | Some s -> Shrink.list ~shrink:s in
+    let ac = QCheck.make ~shrink:(Shrink.filter (cmds_ok Spec.init_state) shrinker) cmds_gen in
+    (match (Spec.arb_cmd s).print with
+     | None   -> ac
+     | Some p -> set_print (Print.list p) ac)
+  (** A generator of command sequences. Accepts the initial state as parameter. *)
+
+  let rec interp_agree s sut cs = match cs with
+    | [] -> true
+    | c::cs ->
+      let res = Spec.run c sut in
+      let b   = Spec.postcond c s res in
+      let s'  = Spec.next_state c s in
+      b && interp_agree s' sut cs
+  (** Checks agreement between the model and the system under test
+      (stops early, thanks to short-circuit Boolean evaluation). *)
+
+  let rec check_disagree s sut cs = match cs with
+    | [] -> None
+    | c::cs ->
+      let res = Spec.run c sut in
+      let b   = Spec.postcond c s res in
+      let s'  = Spec.next_state c s in
+      if b
+      then
+        match check_disagree s' sut cs with
+        | None -> None
+        | Some rest -> Some ((c,res)::rest)
+      else Some [c,res]
+
+  let print_seq_trace trace =
+    List.fold_left
+      (fun acc (c,r) -> Printf.sprintf "%s\n   %s : %s" acc (Spec.show_cmd c) (show_res r))
+      "" trace
+
+  (* ****************************** additions from here ****************************** *)
+
+  let check_and_next (c,res) s =
+    let b  = Spec.postcond c s res in
+    let s' = Spec.next_state c s in
+    b,s'
+
+  (* operate over arrays to avoid needless allocation underway *)
+  let interp_sut_res sut cs =
```

Comment:
A "default interpreter" feels a bit overkill when there are only two "backend interpreters" :thinking: 
The alternative to including a "default" interpreter targeting `Domain` and overriding it in `STM_Thread` would be to let each of `STM_Domain` and `STM_Thread` provide their own. 

Can you share your thoughts on this choice? (I might be missing something)

Reply:
I agree this is not the best design. I was focusing on having `agree_test_conc` finding the expected bugs (which still fail).
 
There is also the re-exportation of four functions in `STM_Seq` (the reason is that they are needed in `src/lockfree/ws_deque_test.ml`) which is not a great design and is related to this one.